clippy_lints/src/types.rs

```diff
@@ -955,7 +955,16 @@ impl<'tcx> LateLintPass<'tcx> for UnitArg {
                     .iter()
                     .filter(|arg| {
                         if is_unit(cx.typeck_results().expr_ty(arg)) && !is_unit_literal(arg) {
```

Comment:
Can you check the `TypeFlags` to see if the type was resolved from a variable? Maybe `expr_ty(arg).needs_subst()`.

Reply:
Sorry, for the slow response. I missed your comment regarding the type flags. Checking `needs_subst()` does not seem to have the desired effect. The following example passes linting with my implemention, but running `needs_subst()` on the type for the argument of `foo` returns `false`:
```
    let arg = if true {
        1;
    };
    foo(arg);
```
I'm not familiar with the compiler internals. I looked at the definition of the various type flags, but did not really understand many of them. Do you have any suggestions?

Reply:
Thinking about it some more, I guess you meant **type variables** (as in the example in issue #6447). I will test that again later today.

Reply:
Yes I'm hoping we can detect type variables by looking that the [`TypeFlags`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeFlags.html) which is readable through [`TypeFoldable`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/fold/trait.TypeFoldable.html) (that's where needs_subst is). Another method in there that might be useful is `has_projections` which means a type with `<X as Y>::Z`. I am guessing a bit here as I don't completely understand when those different flags apply.

I'm afraid this might be even more complicated - we may have to detect "does the type variable reference a type that is defined outside of the impl block?".

Reply:
I think we need a different approach here. Using the example from issue #6447 I checked a __all__ type flags and they are all `false`. So we might not get any help there.
One problem might be that the original lint looks at the type of the expression passed as the argument and not at the type of the argument from the function definition.
I'll investigate this further, but if you have additional ideas I would be grateful.

Reply:
After thinking on this again, I think we should simply not lint if `matches(arg.kind, ExprKind::Path)`. This will not lint if `arg` is a local variable.

It may be bad style to have a variable with unit as the type, but I think that should be linted upon declaration of the variable (not in this lint).

Reply:
Ok, I changed my implementation to check for `ExprKind::Path`.

Reply:
Sorry to send you for a loop there.